tempodb/compaction_block_selector.go

```diff
@@ -58,168 +58,149 @@ func (sbs *simpleBlockSelector) BlocksToCompact() ([]*backend.BlockMeta, string)
-		bi := twbs.blocklist[i]
-		bj := twbs.blocklist[j]
+		entry := timeWindowBlockEntry{
+			meta: b,
+		}
 
-		wi := twbs.windowForBlock(bi)
-		wj := twbs.windowForBlock(bj)
+		age := currWindow - w
+		if activeWindow <= w {
+			// inside active window.
+			// Group by compaction level and window.
+			// Choose lowest compaction level and most recent windows first.
+			entry.group = fmt.Sprintf("A-%v-%016X", b.CompactionLevel, age)
 
-		if activeWindow <= wi && activeWindow <= wj {
-			// inside active window.  sort by:  compaction lvl -> window -> size
-			//  we should always choose the smallest two blocks whos compaction lvl and windows match
-			if bi.CompactionLevel != bj.CompactionLevel {
-				return bi.CompactionLevel < bj.CompactionLevel
-			}
+			// Within group choose smallest blocks first.
+			entry.order = fmt.Sprintf("%016X", entry.meta.TotalObjects)
 
-			if wi != wj {
-				return wi > wj
-			}
+			entry.hash = fmt.Sprintf("%v-%v-%v", b.TenantID, b.CompactionLevel, w)
 		} else {
-			// outside active window.  sort by: window -> compaction lvl -> size
-			//  we should always choose the most recent two blocks that can be compacted
-			if wi != wj {
-				return wi > wj
-			}
+			// outside active window.
+			// Group by window only.  Choose most recent windows first.
+			entry.group = fmt.Sprintf("B-%016X", age)
 
-			if bi.CompactionLevel != bj.CompactionLevel {
-				return bi.CompactionLevel < bj.CompactionLevel
-			}
+			// Within group chose lowest compaction lvl and smallest blocks first.
+			entry.order = fmt.Sprintf("%v-%016X", b.CompactionLevel, entry.meta.TotalObjects)
+
+			entry.hash = fmt.Sprintf("%v-%v", b.TenantID, w)
 		}
 
-		return bi.TotalObjects < bj.TotalObjects
+		twbs.entries = append(twbs.entries, entry)
+	}
+
+	// sort by group then order
+	sort.SliceStable(twbs.entries, func(i, j int) bool {
+		ei := twbs.entries[i]
+		ej := twbs.entries[j]
+
+		if ei.group == ej.group {
+			return ei.order < ej.order
+		}
+		return ei.group < ej.group
 	})
 
 	return twbs
 }
 
 func (twbs *timeWindowBlockSelector) BlocksToCompact() ([]*backend.BlockMeta, string) {
-	for len(twbs.blocklist) > 0 {
-		// find everything from cursor forward that belongs to this block
-		cursor := 0
-		currentWindow := twbs.windowForBlock(twbs.blocklist[cursor])
-
-		windowBlocks := make([]*backend.BlockMeta, 0)
-		for cursor < len(twbs.blocklist) {
-			currentBlock := twbs.blocklist[cursor]
-
-			if currentWindow != twbs.windowForBlock(currentBlock) {
+	for len(twbs.entries) > 0 {
+		var chosen []timeWindowBlockEntry
+
+		// find everything from cursor forward that belongs to this group
+		// Gather contiguous blocks while staying within limits
+		i := 0
+		for ; i < len(twbs.entries); i++ {
+			for j := i + 1; j < len(twbs.entries); j++ {
+				stripe := twbs.entries[i : j+1]
+				if twbs.entries[i].group == twbs.entries[j].group &&
+					len(stripe) <= twbs.MaxInputBlocks &&
+					totalObjects(stripe) <= twbs.MaxCompactionObjects &&
```

Comment:
does this mean we have to set `max_compaction_objects` to a really high number (so this condition is always true) and then max block size to the value we actually want for it to work?

Reply:
after an offline disucssion agree that this is the behavior and it's fine since this is only a temporary transition period until we drop all blocks that don't have a size.

Reply:
That's correct.  The trace limit is left as a safety guard while Size meta is rolled out.  Existing blocks will not have the Size meta and will be treated as Size 0.  During that time it is unlikely to hit the new size limit if the trace limit is not also increased.